File: JavaChangeDirectorySimulator/src/ChangeDirectory.java
import java.util.ArrayList;

public class ChangeDirectory {
    // ROOT will act as our global root node
    static Directory ROOT;



    // finds the starting directory and then traverses the directories
    static String mycd(String currDir, String newDir) {
        // split new_dir on the character "/"
        String[] newDirArr = newDir.split("/");
        // currSym initialized as null
        String currSym = null;
        // currDirectory initialized as null
        Directory currDirectory = null;

        // get the Directory from the currDir string, catch the error if it doesnt exist
        try {
            currDirectory = getDirectory(currDir, ROOT);
        }
        catch(NoSuchFieldException e) {
            return currDir + " : No such file or directory.";
        }

        // if there is nothing left in the array OR the first item was "/", currDirectory is the ROOT
        if(newDirArr.length == 0 || newDirArr[0].equals("")) {
            currDirectory = ROOT;
        }

        // remove the holes ("") from the array
        newDirArr = removeHolesLst(newDirArr);

        // traverse the rest of the array, example [gh,..,klm,.]
        for(String i : newDirArr) {
            currSym = i;
            // if currSym == .
            if (currSym.equals(".")) {
                ; // do nothing to currDir
            }

            // else if currSym == ..
            else if (currSym.equals("..")) {
                // change currDirectory to the parent of the previous
                currDirectory = currDirectory.parent;
            }

            // currSym == name
            else {
                // check if currSym is alphanumeric
                if (isAlphaNumeric(currSym)) {

                    // now check if it is a child of the currDirectory, catch the error if not
                    try {
                        currDirectory = currDirectory.findChild(currSym);
                    }
                    catch(NoSuchFieldException e) {
                        return currSym + " : No such file or directory.";
                    }
                }

                // if symbol isnt alphanumeric, return No such file string
                else {
                    return currSym + " : No such file or directory.";
                }
            }
        }
        // finally get the string path of the final directory
        return getPath(currDirectory);
    }



    // finds the string path given a Directory, currDir
    public static String getPath(Directory currDir) {
        String result = "";

        // simplest case, if currDir is root, just return "/"
        if(currDir.name.equals("/")) {
            return "/";
        }

        // loop through directories and build on result until you reach the root, then return result
        while(!(currDir.name.equals("/"))) {
            result = "/" + currDir.name + result;
            currDir = currDir.parent;
        }
        return result;
    }



    // remove all of the holes from the list
    public static String[] removeHolesLst(String[] inputList) {
        ArrayList<String> resultArr = new ArrayList<String>();
        // loop through the list
        for(String i : inputList) {
            // this checks if current string is empty (hole)
            if(i.equals("")) {
                // if empty, continue
                continue;
            }
            // if not empty, add to result list
            resultArr.add(i);
        }
        return resultArr.toArray(new String[0]);
    }



    // get the directory given a currDirPath string
    public static Directory getDirectory(String currDirPath, Directory root) throws NoSuchFieldException {
        // start at root
        Directory currDir = root;
        // turn the string into a list with no holes split on "/"
        String[] dirArr = removeHolesLst(currDirPath.split("/"));

        // step through dirArr directories
        for (String i : dirArr) {
            // at each level look for the next child Directory, error is caught in find child and thrown
            currDir = currDir.findChild(i);
        }
        // return the child if it exists
        return currDir;
    }



    // find the string path of the Directory, if it exists
    public static Boolean isAlphaNumeric(String currString) {
        if(currString.length() == 0) {
            return false;
        }
        // turn the input string into a character array
        char[] charArr = currString.toCharArray();
        // loop through the characters, checking if each is alphanumeric
        for(char i : charArr) {
            // if not alphanumeric, return false
            if((!Character.isDigit(i)) && (!Character.isAlphabetic(i))) {
                return false;
            }
        }
        return true;
    }



    // initialize network of directories
    public static void createNetwork() {
        // update the global ROOT
        ROOT = new Directory("/");
        Directory v1 = new Directory("abc", ROOT);
        Directory v2 = new Directory("def", v1);
        Directory v3 = new Directory("klm", v1);
        Directory v4 = new Directory("gh", v1);
        Directory v5 = new Directory("ghi", v2);
    }

    // for my test cases I used the directory structure from the test cases given in the email:
    //             ------> gh
    // root       |
    //   / ----> abc ----> def ----> ghi
    //            |
    //             ------> klm



    // all of these test cases (and many more) are implemented in the TestJunit.java file
    public static void main(String args[]) {
        createNetwork();

        // test cases from email
        System.out.println( ChangeDirectory.mycd("/", "abc") ); // /abc
        System.out.println( ChangeDirectory.mycd("/abc/def", "ghi") ); // /abc/def/ghi
        System.out.println( ChangeDirectory.mycd("/abc/def", "..") ); // /abc
        System.out.println( ChangeDirectory.mycd("/abc/def", "/abc") ); // /abc
        System.out.println( ChangeDirectory.mycd("/abc/def", "/abc/klm") ); // /abc/klm
        System.out.println( ChangeDirectory.mycd("/abc/def", "../..") ); // /
        System.out.println( ChangeDirectory.mycd("/abc/def", "../../..") ); // /
        System.out.println( ChangeDirectory.mycd("/abc/def", ".") ); // /abc/def
        System.out.println( ChangeDirectory.mycd("/abc/def", "//////") ); // /
        System.out.println( ChangeDirectory.mycd("/abc/def", "......") ); //  ....... : No such file or directory.
        System.out.println( ChangeDirectory.mycd("/abc/def", "../gh///../klm/.") ); // /abc/klm

        // some additional test cases
        System.out.println( ChangeDirectory.mycd("/", ".") ); // /
        System.out.println( ChangeDirectory.mycd("/", "/abc") ); // /abc
        System.out.println( ChangeDirectory.mycd("/abc", ".") ); // /abc
        System.out.println( ChangeDirectory.mycd("/def", ".") ); // /def : No such file or directory.
        System.out.println( ChangeDirectory.mycd("/abc/def", "cba") ); //  cba : No such file or directory.
        System.out.println( ChangeDirectory.mycd("/abc/def", "") ); // /abc/klm
        System.out.println( ChangeDirectory.mycd("/", "/") ); // /
        System.out.println( ChangeDirectory.mycd("/", "..///.////abc/../abc/def/.././gh/..//def/ghi") ); // /abc/def/ghi
    }
}

File: PythonChangeDirectorySimulator/change_directory.py
# Directory class represents each hypothetical directory
class Directory:
    # the init method deals with the root node as well as all others
    def __init__(self, a_name, a_parent):
        # all nodes will start with a given name and a 
        self.name = a_name
        self.children = []

        # this branch is for the root directory initialization
        if isinstance(a_parent, type(None)):
            # root is its own parent. This is how we deal with something like "/..", which should be "/"
            self.parent = self

        # this branch is for all other directories
        else:
            # add the parent to the self.parent field
            self.parent = a_parent
            # add current object to parents children list
            a_parent.children.append(self)

    def __eq__ (self, other):
        # first we check if the other is of type None, this would cause problems 
        if isinstance(other, type(None)):
            return False
        # if it is not None, we check the equality of the three fields. We only check the names of parents for the sake of avoiding endless recursion
        return self.name == other.name and self.parent.name == other.parent.name and self.children == other.children

    # this method finds a child in children list if it exists and returns the Directory
    def findChild(self, child_name):
        # loop throw the children list
        for child in self.children:
            # if the name of the child is correct, return it
            if child.name == child_name:
                return child
        # if no children match, return None
        return None



# finds the starting directory and then calls my_cd_helper to traverse the directories
def my_cd(curr_dir_path, new_dir):
    # split new_dir on the character "/"
    new_dir_list = new_dir.split("/")

    # get rid of holes in new_dir_list
    try:
        while True:
            new_dir_list.remove("") 
    except ValueError:
        pass
    
    # check if starting node is a legal path, if not return no such file string
    if(isinstance(getNode(curr_dir_path, global_root), type(None))):
        return curr_dir_path + " : No such file or directory."

    # here there was nothing after splitting and removing holes
    if new_dir_list == []: 
        return "/"
    
    # if first char is /, change currDir to global_root
    if  new_dir[0] == "/": 
        return my_cd_helper(global_root, new_dir_list)

    # recursively call my_cd_helper
    return my_cd_helper(getNode(curr_dir_path, global_root), new_dir_list)



# traverse the curr_dir_node with the new_dir_list recursively
def my_cd_helper(curr_dir_node, new_dir_list):
    # base case, list has been emptied
    if new_dir_list == []:
        return getPath(curr_dir_node)

    # case of ".", nothing should change in current directory
    elif new_dir_list[0] == ".":
        # pop the first item from the list and recurse
        new_dir_list.pop(0)
        return my_cd_helper(curr_dir_node, new_dir_list)

    # case of "..", go to parent directory
    elif new_dir_list[0] == "..":
        # pop the first item from the list and recurse
        new_dir_list.pop(0)
        return my_cd_helper(curr_dir_node.parent, new_dir_list)

    # else we have a string that we need to look closer at
    else:
        # check if we have an alphanumeric string
        if new_dir_list[0].isalnum():
            # now check if the child is in the current directory children list
            curr_dir_node = curr_dir_node.findChild(new_dir_list[0])

            # if it isnt in the children list, return no file string
            if(isinstance(curr_dir_node, type(None))):
                return new_dir_list[0] + " : No such file or directory."

            # otherwise, pop the item form the list and continue
            new_dir_list.pop(0)
            return my_cd_helper(curr_dir_node, new_dir_list) 

        # if the string isnt alphanumeric, it cant be a file path, so return no such file string
        else:
            return new_dir_list[0] + " : No such file or directory."



# get the Directory node that the curr_dir string points to
def getNode(dir_path, root):
    # start at the root node
    curr_dir_node = root 

    # split dir_path string into a list on character "/""
    dir_path_list = dir_path.split("/") 

    # get rid of holes in dir_path_list
    try:
        while True:
            dir_path_list.remove("")
    except ValueError:
        pass

    # traverse dir_path_list
    for new_path in dir_path_list: 
        # on each level look for the child directory with name new_path
        curr_dir_node = curr_dir_node.findChild(new_path)

        # if it doesnt exist, return None
        if(isinstance(curr_dir_node, type(None))):
            return None

    # finally return the node that we were looking for
    return curr_dir_node



# get the path of the Directory node (as a string)
def getPath(dir_node):
    # simplest case, return the root string
    if dir_node.name == "/":
        return "/"

    # start with the name of the current node and build from that
    result = dir_node.name

    # loop until we reach the root and then we will return
    while True:
        # update node with parent node
        dir_node = dir_node.parent
        # if we reach the root, return
        if dir_node.name == "/":
            return "/" + result 
        # otherwise build on result string and continue looping
        result =  dir_node.name + "/" + result



# initialize network of directories
def create_network():
    abc = Directory("abc", global_root)
    def_ = Directory("def", abc)
    gh = Directory("gh", abc)
    klm = Directory("klm", abc)
    ghi = Directory("ghi", def_)

# for my test cases I used the directory structure from the test cases given in the email: 
#             ------> gh
# root       |
#   / ----> abc ----> def ----> ghi
#            |
#             ------> klm



# global variable for the root directory
global_root = Directory("/", None)


# all of these test cases (and many more) are implemented in the test_change_directory.py file
def main():
    create_network()
    
    # test cases from email
    print(my_cd("/", "abc"))                        # /abc
    print(my_cd("/abc/def", "ghi"))                 # /abc/def/ghi
    print(my_cd("/abc/def", ".."))                  # /abc
    print(my_cd("/abc/def", "/abc"))                # /abc
    print(my_cd("/abc/def", "/abc/klm"))            # /abc/klm
    print(my_cd("/abc/def", "../.."))               # /
    print(my_cd("/abc/def", "../../.."))            # /
    print(my_cd("/abc/def", "."))                   # /abc/def
    print(my_cd("/abc/def", "//////"))              # /
    print(my_cd("/abc/def", "......"))              # ...... : No such file or directory.
    print(my_cd("/abc/def", "../gh///../klm/."))    # /abc/klm

    # some additional test cases
    print(my_cd("/abc/def", "cba"))                 # cba : No such file or directory.
    print(my_cd("/", "."))                          # / 
    print(my_cd("/", "/abc"))                       # /abc
    print(my_cd("/abc", "."))                       # /abc
    print(my_cd("/def", "."))                       # def : No such file or directory.
    print(my_cd("/", "/"))                          # / 
    print(my_cd("/", "..///.////abc/../abc/def/.././gh/..//def/ghi")) # /abc/def/ghi



if __name__ == "__main__":
    main()
